src/community/security/oauth2-core/src/main/java/org/geoserver/security/oauth2/GeoServerOAuthAuthenticationFilter.java

```diff
@@ -99,76 +99,70 @@ public AuthenticationEntryPoint getAuthenticationEntryPoint() {
-                httpRequest.getSession(false).invalidate();
+            HttpServletRequest httpRequest = (HttpServletRequest) request;
+
+            /*
+             * This cookie works only locally, when accessing the GeoServer GUI and on the same domain. For remote access you need to logout from the
+             * GeoServer GUI.
+             */
+            final String customSessionCookie = getCustomSessionCookieValue(httpRequest);
+
+            if (accessToken == null && customSessionCookie == null && authentication == null) {
+                final AccessTokenRequest accessTokenRequest =
+                        restTemplate.getOAuth2ClientContext().getAccessTokenRequest();
+                if (accessTokenRequest != null && accessTokenRequest.getStateKey() != null) {
+                    restTemplate
+                            .getOAuth2ClientContext()
+                            .removePreservedState(accessTokenRequest.getStateKey());
+                }
+
                 try {
-                    httpRequest.logout();
-                    authentication = null;
-                } catch (ServletException e) {
-                    LOGGER.fine(e.getLocalizedMessage());
+                    accessTokenRequest.remove("access_token");
+                } finally {
+                    SecurityContextHolder.clearContext();
+                    httpRequest.getSession(false).invalidate();
+                    try {
+                        httpRequest.logout();
+                        authentication = null;
+                    } catch (ServletException e) {
+                        LOGGER.fine(e.getLocalizedMessage());
+                    }
+                    LOGGER.fine("Cleaned out Session Access Token Request!");
                 }
-                LOGGER.fine("Cleaned out Session Access Token Request!");
             }
-        }
 
-        if ((authentication == null && accessToken != null)
-                || (accessToken != null && token == null)
-                || authentication == null
-                || (authentication != null
-                        && authorities.size() == 1
-                        && authorities.contains(GeoServerRole.ANONYMOUS_ROLE))) {
-
-            doAuthenticate((HttpServletRequest) request, (HttpServletResponse) response);
-
-            Authentication postAuthentication =
-                    SecurityContextHolder.getContext().getAuthentication();
-            if (postAuthentication != null) {
-                if (cacheAuthentication(postAuthentication, (HttpServletRequest) request)) {
-                    getSecurityManager()
-                            .getAuthenticationCache()
-                            .put(
-                                    getName(),
-                                    getCacheKey((HttpServletRequest) request),
-                                    postAuthentication);
+            if ((authentication == null && accessToken != null)
```

Comment:
This condition stands basically for `authentication==null || accessToken !=null` so since authentication should always be null the code in the if should always executed.